File: packages/moshe-sdk-python/src/moshe/_lineage.py
from __future__ import annotations

from dataclasses import dataclass

from ._interfaces import ArtifactStore


@dataclass(frozen=True)
class LineageNode:
    path: str
    classification: str
    depth: int
    sources: list[str]


@dataclass(frozen=True)
class LineageReport:
    root: str
    found: bool
    max_depth: int
    nodes: list[LineageNode]


async def resolve_lineage(store: ArtifactStore, root_path: str, max_depth: int = 10) -> LineageReport:
    root_artifact = await store.get_artifact(root_path)
    if root_artifact is None:
        return LineageReport(root=root_path, found=False, max_depth=0, nodes=[])

    nodes: list[LineageNode] = []
    visited: set[str] = set()
    queue: list[tuple[str, int]] = [(root_path, 0)]

    while queue:
        current_path, depth = queue.pop(0)
        if current_path in visited:
            continue
        visited.add(current_path)

        artifact = await store.get_artifact(current_path)
        sources = list(artifact.provenance_chain) if artifact and artifact.provenance_chain else []
        nodes.append(
            LineageNode(
                path=current_path,
                classification=artifact.classification if artifact is not None else "CLEAN",
                depth=depth,
                sources=sources,
            )
        )

        if depth >= max_depth:
            continue

        for source_path in sources:
            if source_path not in visited:
                queue.append((source_path, depth + 1))

    return LineageReport(
        root=root_path,
        found=True,
        max_depth=max((node.depth for node in nodes), default=0),
        nodes=nodes,
    )

File: packages/moshe-core/src/lineage.ts
import type { ArtifactStore } from './interfaces.js';
import type { ArtifactRecord } from './state.js';

export interface LineageNode {
  path: string;
  classification: ArtifactRecord['classification'];
  depth: number;
  sources: string[];
}

export interface LineageReport {
  root: string;
  found: boolean;
  maxDepth: number;
  nodes: LineageNode[];
}

export async function resolveLineage(
  store: ArtifactStore,
  rootPath: string,
  maxDepth = 10
): Promise<LineageReport> {
  const rootArtifact = await store.getArtifact(rootPath);
  if (!rootArtifact) {
    return {
      root: rootPath,
      found: false,
      maxDepth: 0,
      nodes: []
    };
  }

  const nodes: LineageNode[] = [];
  const visited = new Set<string>();
  const queue: Array<{ path: string; depth: number }> = [{ path: rootPath, depth: 0 }];

  while (queue.length > 0) {
    const current = queue.shift()!;
    if (visited.has(current.path)) {
      continue;
    }

    visited.add(current.path);

    const artifact = await store.getArtifact(current.path);
    const sources = artifact?.provenanceChain ?? [];
    nodes.push({
      path: current.path,
      classification: artifact?.classification ?? 'CLEAN',
      depth: current.depth,
      sources
    });

    if (current.depth >= maxDepth) {
      continue;
    }

    for (const sourcePath of sources) {
      if (!visited.has(sourcePath)) {
        queue.push({ path: sourcePath, depth: current.depth + 1 });
      }
    }
  }

  return {
    root: rootPath,
    found: true,
    maxDepth: nodes.reduce((max, node) => Math.max(max, node.depth), 0),
    nodes
  };
}
